File: python/1011. Capacity To Ship Packages Within D Days.py
class Solution(object):
    def shipWithinDays(self, weights, D):
        """
        :type weights: List[int]
        :type D: int
        :rtype: int
        """
        def canShip(weights, D, K):
            cur = K
            for w in weights:
                if w > K:
                    return false
                if w > cur:
                    cur = K
                    D -= 1
                cur -= w
            return D > 0

        r = sum(weights)
        l = max(weights)
        while l <= r:
            mid = l + (r - l)/2
            if canShip(weights, D, mid):
                r = mid - 1
            else:
                l = mid + 1
        return l

File: go/src/leetcode/1011. Capacity To Ship Packages Within D Days/1011. Capacity To Ship Packages Within D Days.go
package main

import (
	"fmt"
	"math"
)

func shipWithinDays(weights []int, D int) int {
	l, r := max(weights), sum(weights)

	canShip := func(k, day int) bool {
		cur := k
		for _, w := range weights {
			if w > cur {
				cur = k
				day--
			}
			cur -= w
		}
		return day > 0
	}

	var mid int
	for l <= r {
		mid = l + (r-l)>>1
		if canShip(mid, D) {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}
	return l
}

func sum(a []int) int {
	s := 0
	for _, x := range a {
		s += x
	}
	return s
}

func max(a []int) int {
	maxVal := math.MinInt32
	for _, x := range a {
		if x > maxVal {
			maxVal = x
		}
	}
	return maxVal
}

func main() {
	fmt.Println(shipWithinDays([]int{1,2,3,4,5,6,7,8,9,10}, 5))
	fmt.Println(shipWithinDays([]int{3,2,2,4,1,4}, 3))
	fmt.Println(shipWithinDays([]int{1,2,3,1,1}, 4))
}